### ISSUE TITLE
pairs() nondeterminism causes OOS

### ISSUE THREAD

**Issue by Captain-Fwiffo** [Aleph-One-Marathon/alephone#214]
Greetings, fellow scientists.

I recently tried to beat a score Adminn_1 set in his XBLA Survival plugin on the map Hollow Apollo in the peack Magenta Filter 1.5. I succeeded, but to my dismay, the film does not work; the film ends in different ways most times and never plays as it is supposed to. I will link all these later, but for now here's the film.

[hollow apollo 29k.zip](https://github.com/Aleph-One-Marathon/alephone/files/5132822/hollow.apollo.29k.zip)

Katharan in the Discord server tested the film and plugin on the RC1 and encountered the same problem. I just tested it in Aleph One 1.2.1 and it was borked there too.

Some theories were floated about what's causing it to mess up, including the use of random_location in the lua file among others. I wouldn't know anything about that, being a lua virgin.

http://simplici7y.com/items/xbla-survival-2

http://simplici7y.com/items/magenta-filter

**Comment by Captain-Fwiffo** [comment]
After treellama's investigation, I am informed this is an issue with the XBLA Survival script's use of the pairs(). A 🌶️ member of the Discord server informed me it was a matter of fixing a script, not Aleph One.

**Comment by treellama** [comment]
I think this is just going to keep biting scripts until it is fixed. A couple of options:

1. Rewrite pairs() to sort the table behind the scenes, and return variables in sorted order. Downside: slower
2. Pass the game seed into lua_newstate instead of using luai_makeseed (which is seeded based on time). Downside: determinism means less ASLR protection.

Meanwhile, @Hopper262 if I send you an updated XBLA Survival script and map can you replace your entry on S7?

**Comment by treellama** [comment]
Option 3: add a prominent note to the documentation not to use the function
Option 4: remove the pairs() iterator altogether!

**Comment by Hopper262** [comment]
Yeah, it hadn't crossed my mind that pairs() would be different from run to run. (In retrospect, I should have known better; I've used some versions of Perl with similar behavior.)

The pairs function can be useful, so it'd be nice to have a deterministic version in our environment. Option 1 sounds best, but it's also the most work to build and maintain through Lua upgrades -- I don't understand how much work to judge if it's worthwhile. The speed isn't a problem to me, when the alternative is avoiding it altogether and replacing it with even slower hand-rolled Lua. If we really had a use case for speed, I assume we could expose the original as pairs_fast().

I think this version of the Lua script will avoid OOS; anyone care to test it? I can re-merge and update S7 when it's good, or I'll post yours if you have it ready, treellama.

[XBLA_Survival.lua.zip](https://github.com/Aleph-One-Marathon/alephone/files/5138974/XBLA_Survival.lua.zip)

**Comment by Captain-Fwiffo** [comment]
Your new script is borked! It doesn't work! treellama beat you in a fair contest of script writing. Maybe he'll let you look at the code if you supplicate.

**Comment by treellama** [comment]
Haha they are exactly the same except mine is not missing and `end`. Great minds think alike.

[XBLA_Survival.zip](https://github.com/Aleph-One-Marathon/alephone/files/5142070/XBLA_Survival.zip)
